components/brave_new_tab_ui/reducers/grid_sites_reducer.ts

```diff
@@ -27,81 +37,57 @@ export const gridSitesReducer: Reducer<NewTab.GridSitesState | undefined> = (
   const startingState = state
 
   switch (action.type) {
-    case types.GRID_SITES_SET_FIRST_RENDER_DATA: {
-      // Update existing default SR top sites in gridSites.
-      if (payload.defaultSuperReferralTopSites) {
-        state = gridSitesState
-            .gridSitesReducerUpdateDefaultSuperReferralTopSites(state, payload.defaultSuperReferralTopSites)
-      }
-
-      // If there are legacy values from a previous
-      // storage, update first render data with it
-      state = gridSitesState
-        .gridSitesReducerSetFirstRenderDataFromLegacy(
-          state,
-          startingState.legacy
-        )
-      // Now that we stored the legacy reference, delete it
-      // so it won't override gridSites in further updates
-      if (startingState.legacy) {
-        delete startingState.legacy
-      }
-
-      // New profiles just store what comes from Chromium
-      state = gridSitesState
-        .gridSitesReducerSetFirstRenderData(state, payload.topSites)
-
-      // Handle default top sites data only once.
-      if (payload.defaultSuperReferralTopSites && !storage.isDefaultSuperReferralTopSitesAddedToPinnedSites()) {
-        state = gridSitesState
-          .gridSitesReducerSetDefaultSuperReferralTopSites(state, payload.defaultSuperReferralTopSites)
-        storage.setDefaultSuperReferralTopSitesAddedToPinnedSites()
-      }
-
-      // Cached gridSites can be updated when history has modified.
-      state = gridSitesState.gridSitesReducerDataUpdated(state, state.gridSites)
-      break
-    }
-
     case types.GRID_SITES_DATA_UPDATED: {
-      state = gridSitesState
-        .gridSitesReducerDataUpdated(state, payload.gridSites)
-      break
-    }
-
-    case types.GRID_SITES_TOGGLE_SITE_PINNED: {
-      state = gridSitesState
-        .gridSitesReducerToggleSitePinned(state, payload.pinnedSite)
+      const { gridSites, customLinksEnabled, visible } = payload
+      setMostVisitedSettings(customLinksEnabled, visible, false)
```

Comment:
This doesn't seem right.

When any data changes, C++ is firing `most-visited-info-changed` and providing the new data. The handler in apiEventsToStore is then calling this Redux action (`GRID_SITES_DATA_UPDATED`) which should simply be setting that new data on the store. However it's also calling `setMostVisitedSettings` which is an API to tell C++ to save new settings, which seems a bit circular and not what we want to do.

I think we're missing a redux action, something like `GRID_SITES_CHANGE_REQUESTED` or `GRID_SITES_SETTING_SAVE` which would call the external API to make the change request and either leave the redux data update to the action that would be fired from c++ as a result of that API call, or set it pre-emptively as is done here.

Reply:
This has a potential to be circular, yes- the `false` argument means not to call InstantService (which would cause that condition!). I'll push an update here in a min which uses redux instead

Reply:
FIXED